### ISSUE TITLE
resource_aws_iam_instance_profile.go, potentially destroy not interpolating correctly.

### ISSUE THREAD

**Issue by alejandroEsc** [hashicorp/terraform#8719]
When trying to tear down aws iam resources: (profiles, roles, policies) the following errors occur:

```
[31m  * aws_iam_role.kubernetes_master_role: "name" must match [\w+=,.@-][0m[0m
[31m  * aws_iam_role.kubernetes_worker_role: "name" must match [\w+=,.@-][0m[0m
terraform destroy failed with return code 0, retrying...
There are warnings and/or errors related to your configuration. Please
fix these before continuing.
```

Note, the only possible issue was thought to be underscores, however underscores are allowed by aws for the resources in question and regex looks correct?

Additionally, we tried to remove underscores via interpolation (works on create):

```
name = "${replace("${var.aws_user_prefix}-${var.cluster_name}-master-policy", "_", "-")}"
```

however, the same error occurred, which leaves us thinking it may be due to interpolation not working as it should? 

Note, that creation of the resources works just fine.

### Terraform Version
0.6.16 and current

### Affected Resource(s)
- aws iam roles 
- aws iam policies
- aws iam profiles

If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

### Terraform Configuration Files
```
resource "aws_iam_role" "kubernetes_master_role" {
  name               = "${replace("${var.aws_user_prefix}-${var.cluster_name}-kubernetes-master-role", "_", "-")}"
  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": "sts:AssumeRole",
      "Principal": {
        "Service": "ec2.amazonaws.com"
      },
      "Effect": "Allow",
      "Sid": ""
    }
  ]
}
EOF
}
```

### Expected Behavior
Destroy should work.

### Actual Behavior
Destroy stops abruptly with the following error:
```
[31m  * aws_iam_role.kubernetes_master_role: "name" must match [\w+=,.@-][0m[0m
[31m  * aws_iam_role.kubernetes_worker_role: "name" must match [\w+=,.@-][0m[0m
terraform destroy failed with return code 0, retrying...
There are warnings and/or errors related to your configuration. Please
fix these before continuing.
```

### Factoid
Creation of the resources works just fine.

**Comment by alejandroEsc** [comment]
closing for now, I want to investigate this further to get the proper circumstances to repro this issue consistently.

**Comment by glenn-sontheimer** [comment]
I have discovered that this error is generated when the name violates an AWS requirement.  This error was generated for me when using an inline-if for an IAM role name and I had appended an extra "}" to the end of the string.  That "}" was included as the last character in the role name but it is not an allowed character for IAM role names.  

What I would to see is for Terraform to help us interpret these kinds of error messages.  At a minimum let us know that the error is being generated by AWS rather than Terraform.
